### PR TITLE
Implement PoS for proposers

### LINKED ISSUE #724: changeCurrentProposer is not being tested

### PR DESCRIPTION
**Pull request by daveroga**
Feat #726
Fixes #724

Implement proposer selection and staking based on the PR #506 that was developed in a feature branch following the proposal #283.

## Features
- Add operations in the contracts for staking ETH. In the future decide if it's ETH or MATIC.
- Add/Update endpoints for this proposers staking in optimist.
- Add/Update operations to SDK for this proposers staking.
- This staking will be the base of the staking power when running spans of the weighted round robin algorithm for proposers selection in next task
- Implement weighted round robin algorithm based on the proposer staking for proposer selection

## Actions
- registerProposer now pass the amount to stake instead of bond. 
- updateProposer could update the URL and increment the stake .
- No REGISTRATION_BOND now but a MINIMUM_STAKE (100 wei for testing now).
- COOLING_OFF_PERIOD moved to CHALLENGE_PERIOD (1 week) for clarity.
- Changed messages from require statements in contracts to best practices used by Openzeppelin. <contract name>: <message>. So you know the contract where the message has been thrown. Useful now when debugging.
`withdrawBond` moved to `withdrawStake` and used in the same way but with the stake of the proposer instead of the bond, so the proposer can do the withdraw of their stake. Previously to that, the proposer should call `deregisterProposer` to stop being proposer and wait this CHALLENGE_PERIOD to be allowed to do the withdraw.
- `TimeLockedBond` moved to `TimeLockedStake` and added `challengeLocked` to store the amount locked by blocks proposed still in CHALLENGE_PERIOD and not claimed yet.
```
    struct TimeLockedStake {
        uint256 amount; // The amount held
        uint256 challengeLocked; // The amount locked by block proposed still in CHALLENGE_PERIOD
        uint256 time; // The time the funds were locked from
    }
```
- `changeCurrentProposer` will be based on this new proposer selection.
- Update test `e2e/protocol/proposer.test.mjs` for testing PoS features. `npm run test-e2e-protocol`.

Regarding issue of testing https://github.com/EYBlockchain/nightfall_3_private/issues/205 for testing the proposer, they were missing tests for:
- [x] Block Propose
- [x] Change proposer
- [x] Request block payment
- [x] Restrictions to add more than boot proposer

## Discuss
- Define first approach with a set of parameter numbers for testnet (minimum stake, proposer count, number of springs in a span, ...) that we should analyze the suitable numbers for testnet/mainnet.

### CODE REVIEW

**Review by israelboudoux — CHANGES_REQUESTED** [review]
Hello @daveroga, would you mind add some unit tests re this change please?
Thank you!

**Comment by daveroga** [comment]
> 

Yes, I explained in the description of the PR that tests were added in `npm run test-e2e-protocol`

**Comment by israelboudoux** [comment]
> > Hello @daveroga, would you mind add some unit tests re this change please?
> > Thank you!
> 
> Yes, I explained this in the description of the PR that tests were added in `npm run test-e2e-protocol` to test the functionality of the PoS.

I know my friend, but **e2e** tests _per se_ aren't better than **unit tests**, and we agreed that we will improve this for every change we do. All I'm asking is for adding some unit tests, it should not be so many, but enough to cover the most important bits. If we do not start doing this now, we will never do it and our codebase will continue being unreliable and we will continue spending a lot of time debugging when an e2e test fails.

Thank you!

**Comment by daveroga** [comment]
> > > Hello @daveroga, would you mind add some unit tests re this change please?
> > > Thank you!
> > 
> > 
> > Yes, I explained this in the description of the PR that tests were added in `npm run test-e2e-protocol` to test the functionality of the PoS.
> 
> I know my friend, but **e2e** tests _per se_ aren't better than **unit tests**, and we agreed that we will improve this for every change we do. All I'm asking is for adding some unit tests, it should not be so many, but enough to cover the most important bits. If we do not start doing this now, we will never do it and our codebase will continue being unreliable and we will continue spending a lot of time debugging when an e2e test fails.
> 
> Thank you!

Discussing in meeting with the team it's better to wait for the hardhat migrated version in order to create the unit tests.

**Review by israelboudoux — APPROVED** [review]

File: cli/lib/nf3.mjs [hunk 1]
```diff
@@ -813,6 +814,35 @@ class Nf3 {
     });
   }
 
+  /**
+    Get all the list of existing proposers.
```
Comment (druiz0992) [inline cli/lib/nf3.mjs:818]: This comment seems incorrect

Comment (daveroga) [reply cli/lib/nf3.mjs:818]: Yes, updated

File: nightfall-deployer/contracts/Config.sol [hunk 2]
```diff
@@ -6,14 +6,17 @@ import './Ownable.sol';
 import './Structures.sol';
 
 contract Config is Ownable, Structures {
-    uint256 constant REGISTRATION_BOND = 10 wei; // TODO owner can update
+    uint256 constant MINIMUM_STAKE = 100 wei; // TODO owner can update
```
Comment (druiz0992) [inline nightfall-deployer/contracts/Config.sol:9]: for a future PR, these parameters need to be configurable

**Review by druiz0992 — CHANGES_REQUESTED** [review]
Minor changes requested

File: nightfall-deployer/contracts/Config.sol (hunk 2 above)
Comment (daveroga) [reply nightfall-deployer/contracts/Config.sol:9]: Yes, defined in issue #1018

**Review by druiz0992 — APPROVED** [review]

**Pull request merged**